File: backend/services/draft_safety.py
"""
Draft-safety guard — Python mirror of `frontend/src/lib/draftSafety.js`.

Prevents raw AI prompts, unfilled template placeholders, or empty machine
output from ever being treated as a ready-to-send outreach message. Used
by the audit endpoint that scans persisted Airtable message fields for
records where the composer would have produced garbage.

Keep patterns in sync with the JS version. Each pattern must be a strong
signal — false positives would flag legitimate outreach.
"""
from __future__ import annotations
import re
from typing import Optional, Tuple

# Prompt-shaped openings. Anchored to start of trimmed text.
_PROMPT_OPENERS = [
    re.compile(r"^you are (a |an |the )", re.I),
    re.compile(r"^you're (a |an |the )", re.I),
    re.compile(
        r"^please (enrich|pull|research|provide|generate|summarize|extract|analyze|classify|score)\b",
        re.I,
    ),
    re.compile(r"^i am (an? )?(ai|assistant|language model|claude|chatbot)\b", re.I),
    re.compile(r"^as an? (ai|assistant|language model)\b", re.I),
    re.compile(r"^\[(system|assistant|user|human|instructions?|prompt|role|task)\]", re.I),
    re.compile(
        r"^(system|assistant|user|human|instructions?|prompt|role|task)\s*[:—-]\s",
        re.I,
    ),
    re.compile(r"^###\s*(instructions?|role|task|context|prompt|system)", re.I),
    # Directive-to-operator openers — bare imperatives Claude uses to tell
    # Ryan what to do next. Live prod bug (2026-02-19) was "Build a short
    # finish-fit memo…" stored in `current_recommendation` and piped into
    # the mailto body.
    re.compile(
        r"^(build|map|monitor|verify|review|prepare|investigate|extract|scrape|"
        r"classify|tag|categorize|assess|score|rate|log)\s+"
        r"(a|an|the|it|this|that|these|those|recurring|any|all|for|from|to|through)\b",
        re.I,
    ),
    re.compile(r"^(wait for|check for|update|ensure|confirm) (a |the |any |all )", re.I),
]

# Structural markers anywhere in the text.
_STRUCTURAL_MARKERS = [
    re.compile(r"\{\{[^}]{1,80}\}\}"),
    re.compile(r"<<[A-Z_ ]{3,}>>"),
    re.compile(r"\bTODO\s*:\s*(fill|write|generate|complete|replace)", re.I),
    re.compile(r"\[(FILL|INSERT|REPLACE|TODO)\s*[^\]]*\]", re.I),
    re.compile(r"\b(as an ai|as a language model|i'?m claude|i am claude)\b", re.I),
    re.compile(r"\bopenai\s+(gpt|api)", re.I),
]

_MIN_WORD_CHARS = 3


def looks_like_ai_prompt(text: Optional[str]) -> Tuple[bool, Optional[str]]:
    """Returns (trip, reason). trip=True means the text is NOT safe to
    present as a ready-to-send draft."""
    if text is None:
        return True, "empty"
    s = str(text).strip()
    if not s:
        return True, "empty"
    word_chars = re.sub(r"[^A-Za-z0-9]", "", s)
    if len(word_chars) < _MIN_WORD_CHARS:
        return True, "too_short"
    for pat in _PROMPT_OPENERS:
        if pat.search(s):
            return True, "prompt_opener"
    for pat in _STRUCTURAL_MARKERS:
        if pat.search(s):
            return True, "structural_marker"
    return False, None

File: frontend/src/lib/draftSafety.js
/**
 * Draft-safety guard — prevents raw AI prompts, unfilled template
 * placeholders, or empty machine output from ever reaching the send screen.
 *
 * These heuristics are conservative but explicit: if any of them trips,
 * OpenInMessages HIDES the mailto button and shows a visible error instead
 * of a "ready to send" draft. The trip must be a strong signal — false
 * positives block Ryan from mailing a real lead — so each pattern targets
 * text that a business recipient would never see in an actual outreach.
 *
 * A Python mirror lives in `backend/services/draft_safety.py`. Keep them in
 * sync when adding patterns.
 */

// Prompt-shaped openings. Anchored to the START of trimmed text so a normal
// sentence that happens to contain "You are" mid-body does not trip.
const PROMPT_OPENERS = [
  /^you are (a |an |the )/i,               // "You are a lead-research assistant..."
  /^you'?re (a |an |the )/i,
  /^please (enrich|pull|research|provide|generate|summarize|extract|analyze|classify|score)\b/i,
  /^i am (an? )?(ai|assistant|language model|claude|chatbot)\b/i,
  /^as an? (ai|assistant|language model)\b/i,
  /^\[(system|assistant|user|human|instructions?|prompt|role|task)\]/i,
  /^(system|assistant|user|human|instructions?|prompt|role|task)\s*[:—-]\s/i,
  /^###\s*(instructions?|role|task|context|prompt|system)/i,
];

// Structural markers anywhere in the body. Any single hit is a hard fail.
const STRUCTURAL_MARKERS = [
  /\{\{[^}]{1,80}\}\}/,                    // {{unfilled_placeholder}}
  /<<[A-Z_ ]{3,}>>/,                       // <<PLACEHOLDER>>
  /\bTODO\s*:\s*(fill|write|generate|complete|replace)/i,
  /\[(FILL|INSERT|REPLACE|TODO)\s*[^\]]*\]/i,   // [FILL IN MESSAGE]
  /\b(as an ai|as a language model|i'?m claude|i am claude)\b/i,
  /\bopenai\s+(gpt|api)/i,
];

// Empty / whitespace only. Body must have at least one real word char sequence
// of length 3 to be considered "has content". Trims quoted signatures.
const _MIN_WORD_CHARS = 3;

/**
 * looksLikeAIPrompt — returns { trip: boolean, reason: string|null }.
 * `text` is examined trimmed and case-insensitively for structural markers.
 * A "trip" means the text should NEVER be presented as a ready-to-send draft.
 */
export const looksLikeAIPrompt = (text) => {
  if (text == null) return { trip: true, reason: "empty" };
  const s = String(text).trim();
  if (!s) return { trip: true, reason: "empty" };
  // "Real content" floor — a draft with fewer than N alphanumeric chars is
  // almost certainly a stub. Signature-only bodies are handled by the caller
  // which only appends the signature AFTER this check.
  const wordChars = s.replace(/[^A-Za-z0-9]/g, "");
  if (wordChars.length < _MIN_WORD_CHARS) return { trip: true, reason: "too_short" };

  for (const re of PROMPT_OPENERS) {
    if (re.test(s)) return { trip: true, reason: "prompt_opener" };
  }
  for (const re of STRUCTURAL_MARKERS) {
    if (re.test(s)) return { trip: true, reason: "structural_marker" };
  }
  return { trip: false, reason: null };
};

/**
 * safeDraftBody — returns the trimmed body when it passes the guard, or
 * null when it trips. Callers should treat null as "do not render Send".
 */
export const safeDraftBody = (text) => {
  const { trip } = looksLikeAIPrompt(text);
  return trip ? null : String(text).trim();
};
